cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,213 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.route.store;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.common.io.Codec;
+import co.cask.cdap.common.service.ServiceDiscoverable;
+import co.cask.cdap.common.zookeeper.ZKExtOperations;
+import co.cask.cdap.internal.guava.reflect.TypeToken;
+import co.cask.cdap.proto.id.ProgramId;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import org.apache.twill.common.Threads;
+import org.apache.twill.zookeeper.NodeData;
+import org.apache.twill.zookeeper.OperationFuture;
+import org.apache.twill.zookeeper.ZKClient;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * RouteStore where the routes are stored in a ZK persistent node. This is intended for use in distributed mode.
+ */
+public class ZKRouteStore implements RouteStore {
+  private static final Logger LOG = LoggerFactory.getLogger(ZKRouteStore.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_INTEGER_TYPE = new TypeToken<Map<String, Integer>>() { }.getType();
+  private static final int ZK_TIMEOUT_SECS = 5;
+
+  private final ZKClient zkClient;
+  private final ConcurrentMap<ProgramId, SettableFuture<RouteConfig>> routeConfigMap;
+
+  @Inject
+  public ZKRouteStore(ZKClient zkClient) {
+    this.zkClient = zkClient;
+    this.routeConfigMap = Maps.newConcurrentMap();
+  }
+
+  @Override
+  public void store(final ProgramId serviceId, final RouteConfig routeConfig) {
+    Supplier<RouteConfig> supplier = Suppliers.ofInstance(routeConfig);
+    final SettableFuture<RouteConfig> settableFuture = SettableFuture.create();
+    Futures.addCallback(ZKExtOperations.createOrSet(zkClient, getZKPath(serviceId), supplier, ROUTE_CONFIG_CODEC, 10),
+                        new FutureCallback<RouteConfig>() {
+                          @Override
+                          public void onSuccess(RouteConfig result) {
+                            settableFuture.set(result);
+                            routeConfigMap.put(serviceId, settableFuture);
```

Comment:
Yes, that is right. But if we can a request to fetch config, immediately after deleting it, would it possible that the routeConfigMap will still have the key and thus return the config (before we can a callback on the watch from ZK which will trigger the deletion of the key from the map) ? Is there a problem storing, removing it from routeConfigMap?